File: Leetcode/add_binary.py


class Solution:
	def addBinary(self, a, b):
		sum = ''
		carry = '0'
		ia = len(a) - 1
		ib = len(b) - 1

		while (ib >= 0 or ia >= 0):
			if (ia >= 0 and ib >= 0):
				r, u = self.add(a[ia], b[ib])
				carry, u = self.add(u, carry)
				if r == '1':
					carry = '1'
			elif ia < 0 :
				carry, u = self.add(carry, b[ib])

			elif ib < 0:
				carry, u = self.add(carry, a[ia])
			sum = u + sum
			ib = ib - 1
			ia = ia - 1

		if carry == '1':
			return carry + sum
		return sum

	def add(self, x, y):
		if x == '1' and y == '1':
			return '1' , '0'
		elif x == '0' and y == '0':
			return '0' , '0'

		return '0', '1'

s = Solution()

File: Leetcode/add_binary.c
#include <stdio.h>

char* addBinary(char* a, char* b){
	int ia = 0, ib = 0;

	char *t;
	while (a[ia] || b[ib]){
		if (a[ia]) ia++;
		if (b[ib]) ib++;
	}

	if (ia > ib) {
		t = b;
		b = a;
		a = t;
		ia = ia + ib;
		ib = ia - ib;
		ia = ia - ib;
	}

	char carry = '0';

	while (ia >= 0 || ib >= 0){
		ia--;ib--;
		if (ia > 0 && ib > 0){
			add(&a[ia], &b[ib]);
			add(&carry, &b[ib]);
			if (carry == '1' || a[ia] == '1'){
				carry = '1';
			}
		}else{
			add(&carry, &b[ib]);
		}
	}return b;

}

void add(char *x, char *y){
	if (*x == '1' && *y == '1'){
		*x = '1'; 
		*y = '0';
	}else if (*x == '0' && *y == '0'){
		*x = '0';
		*y = '0';
	}else{
		*x = '0';
		*y = '1';
	}
}


int main(){
	char x[10] = "1";
	char y[10] = "1";
	printf("%s\n", addBinary(x, y));
	return 0;
}	


